internal/db/view.go

```diff
@@ -88,3 +93,218 @@ func (db *db) addView(
+	txn := mustGetContextTxn(ctx)
+	identity := GetContextIdentity(ctx)
+
+	p := planner.New(ctx, identity, db.acp, db, txn)
+
+	// temporarily disable the cache in order to query without using it
+	col.Description.IsMaterialized = false
+	col.Description, err = description.SaveCollection(ctx, txn, col.Description)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		var defErr error
+		col.Description.IsMaterialized = true
+		col.Description, defErr = description.SaveCollection(ctx, txn, col.Description)
+		if err == nil {
+			// Do not overwrite the original error if there is one, defErr is probably an artifact of the original
+			// failue and can be discarded.
+			err = defErr
+		}
+	}()
+
+	request, err := db.generateMaximalSelectFromCollection(ctx, col, immutable.None[string](), map[string]struct{}{})
+	if err != nil {
+		return err
+	}
+
+	source, err := p.MakeSelectionPlan(request)
+	if err != nil {
+		return err
+	}
+
+	err = source.Init()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		defErr := source.Close()
+		if err == nil {
+			// Do not overwrite the original error if there is one, defErr is probably an artifact of the original
+			// failue and can be discarded.
+			err = defErr
+		}
+	}()
+
+	err = source.Start()
+	if err != nil {
+		return err
+	}
+
+	hasValue, err := source.Next()
+	if err != nil {
+		return err
+	}
+
+	// View items are currently keyed by their index, starting at 1.
+	// The order in which results are returned must be consistent with the results of the
+	// underlying query/transform.
+	var itemID uint
+	for itemID = 1; hasValue; itemID++ {
+		doc := source.Value()
+
+		serializedItem, err := core.MarshalViewItem(doc)
+		if err != nil {
+			return err
+		}
+
+		itemKey := core.NewViewCacheKey(col.Description.RootID, itemID)
+		err = txn.Datastore().Put(ctx, itemKey.ToDS(), serializedItem)
+		if err != nil {
+			return err
+		}
+
+		hasValue, err = source.Next()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (db *db) clearViewCache(ctx context.Context, col client.CollectionDefinition) error {
+	txn := mustGetContextTxn(ctx)
+	prefix := core.NewViewCacheColPrefix(col.Description.RootID)
+
+	q, err := txn.Datastore().Query(ctx, query.Query{
+		Prefix:   prefix.ToString(),
+		KeysOnly: true,
+	})
+	if err != nil {
+		return err
+	}
+
+	for res := range q.Next() {
+		if res.Error != nil {
+			return errors.Join(res.Error, q.Close())
+		}
+
+		err = txn.Datastore().Delete(ctx, ds.NewKey(res.Key))
+		if err != nil {
+			return errors.Join(err, q.Close())
+		}
+	}
+
+	return q.Close()
+}
+
+func (db *db) generateMaximalSelectFromCollection(
+	ctx context.Context,
+	col client.CollectionDefinition,
+	fieldName immutable.Option[string],
+	typesHit map[string]struct{},
+) (*request.Select, error) {
+	// `__-` is an impossible field name prefix, so we can safely concat using it as a separator without risk
```

Comment:
Question: curious were this seperater comes from is a one char or a more symmetric separator not possible?

Reply:
a double underscore prefix is reserved for special gql fields like `__type`, and no special field starts with `-`, thus `__-` is safe(and a bit more obvious than `__y`)